src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ExponentialHistogramAggregator.cs

```diff
@@ -0,0 +1,108 @@
+        public override AggregationStatistics? Collect()
+        {
+            int[][] counters;
+            int count;
+            lock (this)
+            {
+                counters = _counters;
+                count = _count;
+                _counters = new int[ExponentArraySize][];
+                _count = 0;
+            }
+            QuantileValue[] quantiles = new QuantileValue[_config.Percentiles.Length];
+            int nextPercentileIdx = 0;
+            int cur = 0;
+            int target = Math.Max(1, (int)(_config.Percentiles[nextPercentileIdx] * count / 100.0));
+            for (int exponent = 0; exponent < ExponentArraySize; exponent++)
+            {
+                int[] mantissaCounts = counters[exponent];
+                if (mantissaCounts == null)
+                {
+                    continue;
+                }
+                for (int mantissa = 0; mantissa < _mantissaMax; mantissa++)
+                {
+                    cur += mantissaCounts[mantissa];
+                    while (cur >= target)
+                    {
+                        quantiles[nextPercentileIdx] = new QuantileValue(
+                            _config.Percentiles[nextPercentileIdx] / 100.0,
+                            GetBucketCanonicalValue(exponent, mantissa));
+                        nextPercentileIdx++;
+                        if (nextPercentileIdx == _config.Percentiles.Length)
+                        {
+                            return new HistogramStatistics(MeasurementAggregation, quantiles);
+                        }
+                        target = Math.Max(1, (int)(_config.Percentiles[nextPercentileIdx] * count / 100.0));
+                    }
+                }
+            }
+
+            Debug.Assert(_count == 0);
+#pragma warning disable CA1825 // Array.Empty<T>() doesn't exist in all configurations
+            return new HistogramStatistics(MeasurementAggregation, new QuantileValue[0]);
+#pragma warning restore CA1825
+        }
+
+        public override void Update(double measurement)
+        {
+            lock (this)
+            {
+                ref long bits = ref Unsafe.As<double, long>(ref measurement);
+                int exponent = (int)(bits >> ExponentShift);
+                int mantissa = (int)(bits >> _mantissaShift) & _mantissaMask;
+                ref int[] mantissaCounts = ref _counters[exponent];
+                mantissaCounts ??= new int[_mantissaMax];
+                mantissaCounts[mantissa]++;
+                _count++;
+            }
+        }
+
+        // This is the upper bound for negative valued buckets and the
+        // lower bound for positive valued buckets
+        private double GetBucketCanonicalValue(int exponent, int mantissa)
+        {
+            double result = 0;
+            ref long bits = ref Unsafe.As<double, long>(ref result);
+            bits = ((long)exponent << ExponentShift) | ((long)mantissa << _mantissaShift);
+            return result;
```

Comment:
Is this swapped?
In L103 `bits` is `0` as `result = 0` in L102, then `bits` is assigned in L104 to a value. `result` at return on L105 is still `0`.

```suggestion
            long bits = ((long)exponent << ExponentShift) | ((long)mantissa << _mantissaShift);
            return BitConverter.Int64BitsToDouble(bits);
```
(but I'm not sure if I understood the intention here correct).

Reply:
The ref is like creating a pointer in C so any write to bits also changes the value of result. However the BitConverter approach you suggested is more clear so I'm going with that instead👍